### PR TITLE
Command prompt wrap

### LINKED ISSUE #2079: command-prompt doesn't wrap command output
If the command you run happens to output long lines, they will run off the right side of the screen and are unable to be read.

We should implement output text wrapping in the plugin here: https://github.com/DFHack/dfhack/blob/develop/plugins/command-prompt.cpp#L137

### PR DESCRIPTION
**Pull request by softmoth**
Adding word wrapping to the command-prompt responses is simple. But we want to preserve whitespace, which our existing `word_wrap()` (both C++ and Lua) don't do. So here's an upgrade.

Now that word wrapping is in place, it really cries out for scrolling the results up and down. That can be done in a separate PR, though.

Closes #2079

Tests for word_wrap(), to be added once unit testing for C++ is sorted:

```cpp
#include "Internal.h"
#include "MiscUtils.h"

#include <iostream>
#include <string>
using namespace std;

bool compare_result(const vector<string> &expect, const vector<string> &result)
{
    if (result == expect)
    {
        return true;
    }
    else
    {
        auto e = expect.begin();
        auto r = result.begin();
        cout << "# " << setw(20) << left << "Expected" << " " << left << "Got\n";
        while (e < expect.end() || r < result.end())
        {
            cout
                << "# "
                << setw(20) << left << ":" + (e < expect.end() ? *e++ : "") + ":"
                << " "
                << setw(20) << left << ":" + (r < result.end() ? *r++ : "") + ":"
                << "\n";
        }
        return false;
    }
}

int main()
{
    const int plan = 14;

    int test_num = 0;
    int fails = 0;
#define TEST_WORDWRAP(label, args, expect) \
    { \
        ++test_num; \
        vector<string> RESULT; \
        word_wrap args; \
        bool ok = compare_result(expect, RESULT); \
        cout << setw(6) << left << (ok ? "ok" : "not ok") \
                << " " << setw(5) << test_num << " - " << label \
                << "\n"; \
    }

    cout << "1.." << plan << "\n";

    TEST_WORDWRAP("Short line, no wrap", (&RESULT, "12345", 15), vector<string>({"12345"}));

    TEST_WORDWRAP("One long word, wrap", (&RESULT, "12345", 3), vector<string>({"123", "45"}));

    TEST_WORDWRAP("Lines preserved", (&RESULT, "123\n45\n", 15), vector<string>({"123", "45"}));

    TEST_WORDWRAP("Internal spaces per line", (&RESULT, "123  ab cde\n45 x  y   z\n", 15), vector<string>({"123  ab cde", "45 x  y   z"}));

    TEST_WORDWRAP("Internal spaces collapsed", (&RESULT, "123  ab cde\n45 x  y   z\n", 15, true), vector<string>({"123 ab cde", "45 x y z"}));

    TEST_WORDWRAP("Trailing spaces", (&RESULT, "123 \n456  \n789   \n", 15), vector<string>({"123 ", "456  ", "789   "}));

    TEST_WORDWRAP("Trailing spaces collapsed", (&RESULT, "123 \n456  \n789   \n", 15, true), vector<string>({"123 ", "456 ", "789 "}));

    TEST_WORDWRAP("Initial spaces", (&RESULT, " 123\n  456\n   789\n", 15), vector<string>({" 123", "  456", "   789"}));

    TEST_WORDWRAP("Initial spaces collapsed", (&RESULT, " 123\n  456\n   789\n", 15, true), vector<string>({"123", "456", "789"}));

    TEST_WORDWRAP("Basic wrap", (&RESULT, "this is a long line that will wrap, and it has normal spaces in it", 15), vector<string>({"this is a long", "line that will", "wrap, and it", "has normal", "spaces in it"}));

    TEST_WORDWRAP("Wrap with internal spaces", (&RESULT, "this  is a   long line    that     will  wrap,   and    it has   extra   spaces in   it", 15), vector<string>({"this  is a  ", "long line   ", "that     will ", "wrap,   and   ", "it has   extra ", " spaces in   it"}));

    TEST_WORDWRAP("Wrap with internal spaces collapsed", (&RESULT, "this  is a   long line    that     will  wrap,   and    it has   extra   spaces in   it", 15, true), vector<string>({"this is a long", "line that will", "wrap, and it", "has extra", "spaces in it"}));

    TEST_WORDWRAP("One long word, initial spaces, wrap", (&RESULT, "   12345   ", 3), vector<string>({"  ", "123", "45 ", " "}));

    TEST_WORDWRAP("One long word, initial spaces, wrap, collapse", (&RESULT, "   12345   ", 3, true), vector<string>({"123", "45 "}));

    return fails == 0 ? 0 : 1;
}
```

### CODE REVIEW

File: library/MiscUtils.cpp [hunk 1]
```diff
@@ -168,32 +168,58 @@ std::string to_search_normalized(const std::string &str)
...
+bool word_wrap(std::vector<std::string> *out, const std::string &str,
+               size_t line_length, bool collapse_whitespace)
 {
-    out->clear();
-    std::istringstream input(str);
-    std::string out_line;
-    std::string word;
-    if (input >> word)
+    if (line_length == 0)
+        line_length = -1;
```
Comment (lethosor) [inline library/MiscUtils.cpp:176]: I would suggest `SIZE_MAX` as a slightly more clear alternative to the underflow of -1 here

**Review by lethosor — APPROVED** [review]
Looks good! Here's the Lua snippet I used to test, which should cover most of the cases in an 80-character-wide window.

```
print(string.rep('a', 80) .. string.rep('b', 10))
print('-----')
print(string.rep('a', 80) .. string.rep('b', 80) .. string.rep('c', 10))
print('-----')
print(string.rep('a', 78) .. '     ' .. 'bbb')
print('-----')
print(string.rep('a', 78) .. ' ' .. 'bbb')
print('-----')
print('d e  f   g')
```

**Review by lethosor — COMMENTED** [review]
(testing workflow automation)

**Review by lethosor — DISMISSED** [review]
(testing again)

File: library/MiscUtils.cpp (hunk 1 above)
Comment (softmoth) [reply library/MiscUtils.cpp]: Thanks, done.

**Review by lethosor — APPROVED** [review]

**Pull request merged**